src/ray/raylet/node_manager.cc

```diff
@@ -524,34 +524,23 @@ void NodeManager::ProcessNewClient(LocalClientConnection &client) {
   client.ProcessMessages();
 }
 
-void NodeManager::DispatchTasks(const std::list<Task> &ready_tasks) {
-  // Return if there are no tasks to schedule.
-  if (ready_tasks.empty()) {
-    return;
-  }
-
-  // Remember ids of the task we need to remove from ready queue.
-  std::unordered_set<TaskID> removed_task_ids = {};
-
-  for (const auto &task : ready_tasks) {
-    const auto &task_resources = task.GetTaskSpecification().GetRequiredResources();
-    if (!local_available_resources_.Contains(task_resources)) {
-      // Not enough local resources for this task right now, skip this task.
-      // TODO(rkn): We should always skip node managers that have 0 CPUs.
-      continue;
-    }
-    // We have enough resources for this task. Assign task.
-    if (AssignTask(task)) {
-      // We were successful in assigning this task on a local worker, so
-      // remember to remove it from ready queue. If for some reason the
-      // scheduling of this task fails later, we will add it back to the
-      // ready queue.
-      removed_task_ids.insert(task.GetTaskSpecification().TaskId());
+void NodeManager::DispatchTasks() {
+  std::unordered_set<TaskID> removed_task_ids;
+  for (auto &it : local_queues_.GetReadyQueue().GetTasksWithResources()) {
+    for (const auto &task_id : it.second) {
```

Comment:
hmm, so this goes deep into the bucket and schedules all tasks for a given chosen resource shape before advancing to another resource shape. I think this can starve other buckets. I would propose to round robin over the buckets and pick one task from each until no bucket can yield a task that can be dispatched.

Reply:
In other words, instead of going DFS on each ResourceSet bucket, I would do BFS across buckets. I think think BFS in this case would be starvation-free, while DFS is not.

Reply:
Again, starvation should be handled in a different PR and there should be a design doc about it.